정형데이터처리 == pandas사용함  
정형데이터는 딥러닝보다 기계학습이 낫다.

정형데이터 == db에 들어가는 데이터

tensorflow로 ML를 GPU로 돌릴 수 있게 해준다.  
scikit 단점 : gpu못 씀 ㄸㄹㄹ

scikit은 놀랍게도 numpy기반임  
단지 pandas로 scikit learn을 쓰는 방법을 배움...!  
나머지는 pandas기반임  

# [statsmodels](https://www.statsmodels.org/dev/index.html)은 pandas기반으로 통계적 예측분석한다. (중고급 통계분석)
기능 : regression, 시계열분석, 통계분석(pandas에 없는 분석)

통계분석의 관점들   
scikit은 모델성능 높이는데 초점  
statsmodels는 데이터의 해석설명에 초점을 둠 == 그래프그리고...  

R방식과 numpy 선형대수 이용하는 방식 2가지 모두 지원해줌

In [2]:
!pip3 install statsmodels

     |████████████████████████████████| 9.6MB 237kB/s eta 0:00:01
     |████████████████████████████████| 235kB 9.2MB/s eta 0:00:01


In [ ]:
R특징
# 패키지마다 연습용 데이터 존재
# data를 위한 패키지가 따로 있다. scikit보다 더 많이..!

In [4]:
import statsmodels.api as sm  # api는 R관례상 붙음

In [5]:
dir(sm)

['BayesGaussMI',
 'BinomialBayesMixedGLM',
 'Factor',
 'GEE',
 'GLM',
 'GLS',
 'GLSAR',
 'GeneralizedPoisson',
 'Logit',
 'MANOVA',
 'MI',
 'MICE',
 'MICEData',
 'MNLogit',
 'MixedLM',
 'NegativeBinomial',
 'NegativeBinomialP',
 'NominalGEE',
 'OLS',
 'OrdinalGEE',
 'PCA',
 'PHReg',
 'Poisson',
 'PoissonBayesMixedGLM',
 'ProbPlot',
 'Probit',
 'QuantReg',
 'RLM',
 'RecursiveLS',
 'SurvfuncRight',
 'WLS',
 'ZeroInflatedGeneralizedPoisson',
 'ZeroInflatedNegativeBinomialP',
 'ZeroInflatedPoisson',
 '__builtins__',
 '__cached__',
 '__doc__',
 '__file__',
 '__loader__',
 '__name__',
 '__package__',
 '__spec__',
 'add_constant',
 'categorical',
 'cov_struct',
 'datasets',
 'distributions',
 'duration',
 'emplike',
 'families',
 'formula',
 'genmod',
 'graphics',
 'iolib',
 'load',
 'multivariate',
 'nonparametric',
 'qqline',
 'qqplot',
 'qqplot_2samples',
 'regression',
 'robust',
 'show_versions',
 'stats',
 'test',
 'tools',
 'tsa',
 'version',
 'webdoc']

In [6]:
sm.datasets.cancer.DESCRLONG

'The number of breast cancer observances in various counties'

In [7]:
dat = sm.datasets.get_rdataset("Guerry", 'HistData').data # .data하면 pandas format이 된다.
# cache기법으로 저장 가능, 매 번 인터넷에서 안 불러와도 됨

In [8]:
dat

dept Region           Department  Crime_pers  Crime_prop  Literacy  \
0      1      E                  Ain       28870       15890        37   
1      2      N                Aisne       26226        5521        51   
2      3      C               Allier       26747        7925        13   
3      4      E         Basses-Alpes       12935        7289        46   
4      5      E         Hautes-Alpes       17488        8174        69   
5      7      S              Ardeche        9474       10263        27   
6      8      N             Ardennes       35203        8847        67   
7      9      S               Ariege        6173        9597        18   
8     10      E                 Aube       19602        4086        59   
9     11      S                 Aude       15647       10431        34   
10    12      S              Aveyron        8236        6731        31   
11    13      S     Bouches-du-Rhone       13409        5291        38   
12    14      N             Calvados       17577        4500        52   
13    15      C               Cantal       18070       11645        31   
14    16      W             Charente       24964       13018        36   
15    17      W  Charente-Inferieure       18712        5357        39   
16    18      C                 Cher       21934       10503        13   
17    19      C              Correze       15262       12949        12   
18    21      E            Cote-d'Or       32256        9159        60   
19    22      W        Cotes-du-Nord       28607        7050        16   
20    23      C               Creuse       37014       20235        23   
21    24      W             Dordogne       21585       10237        18   
22    25      E                Doubs       11560        5914        73   
23    26      E                Drome       13396        7759        42   
24    27      N                 Eure       14795        4774        51   
25    28      C         Eure-et-Loir       21368        4016        54   
26    29      W            Finistere       29872        6842        15   
27    30      S                 Gard       13115        7990        40   
28    31      S        Haute-Garonne       18642        7204        31   
29    32      S                 Gers       18642       10486        38   
..   ...    ...                  ...         ...         ...       ...   
56    59      N                 Nord       26740        6175        45   
57    60      N                 Oise       28180        6659        54   
58    61      N                 Orne       28329        8248        45   
59    62      N        Pas-de-Calais       23101        4040        49   
60    63      C          Puy-de-Dome       17256       12141        19   
61    64      W      Basses-Pyrenees       16722        8533        47   
62    65      S      Hautes-Pyrenees       12223        9797        53   
63    66      S  Pyrenees-Orientales        6728        7632        31   
64    67      E             Bas-Rhin       12309        4920        62   
65    68      E            Haut-Rhin        7343        4915        71   
66    69      E                Rhone       18793        4504        45   
67    70      E          Haute-Saone       22339        7770        59   
68    71      E       Saone-et-Loire       28391       10708        32   
69    72      C               Sarthe       33913        8294        30   
70    75      N                Seine       13945        1368        71   
71    76      N     Seine-Inferieure       18355        2906        43   
72    77      N       Seine-et-Marne       22201        5786        54   
73    78      N        Seine-et-Oise       12477        3879        56   
74    79      W          Deux-Sevres       18400        6863        41   
75    80      N                Somme       33592        7144        44   
76    81      S                 Tarn       13019        6241        20   
77    82      S      Tarn-et-Garonne       14790        8680        25   
78    83

scikit에는 시계열 분석이 없음  
따라서 statsmodels & facebook prophet사용함

In [11]:
import statsmodels.formula.api as smf  # formula == Formula-compatible models이 있는 패키지
results = smf.ols('Lottery ~ Literacy', data=dat).fit()  # scikit과 다른 방식으로 fit함
# 'Lottery ~ Literacy' : 는 dataframe column

# Lottery가 종속
# Literacy랑 Lottery가 어떤 관계인가?

In [19]:
results.summary()
# 설명 용도로 사용

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                Lottery   R-squared:                       0.133
Model:                            OLS   Adj. R-squared:                  0.123
Method:                 Least Squares   F-statistic:                     12.89
Date:                Tue, 18 Jun 2019   Prob (F-statistic):           0.000555
Time:                        21:54:25   Log-Likelihood:                -392.11
No. Observations:                  86   AIC:                             788.2
Df Residuals:                      84   BIC:                             793.1
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept     64.0896      6.265     10.230      0.000      51.631      76.548
Literacy      -0.5245      0.146     -3.590      0.001      -0.815      -0.234
==============================================================================
Omnibus:                        8.096   Durbin-Watson:                   1.946
Prob(Omnibus):                  0.017   Jarque-Bera (JB):                3.090
Skew:                           0.072   Prob(JB):                        0.213
Kurtosis:                       2.083   Cond. No.                         107.
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

== -0.5245 * Literacy + 64.0896 == Lottery라는 관계가 있다.
intercept == y절편
coef가 기울기

In [21]:
# sklearn으로 하는 방법

In [22]:
from sklearn.linear_model import LinearRegression

In [23]:
lr = LinearRegression()

In [25]:
lr.fit(dat.Literacy.values.reshape(-1,1), dat.Lottery)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None, normalize=False)

In [26]:
vars(lr)

{'fit_intercept': True,
 'normalize': False,
 'copy_X': True,
 'n_jobs': None,
 'coef_': array([-0.52449683]),
 '_residues': 45947.213569626685,
 'rank_': 1,
 'singular_': array([160.08863824]),
 'intercept_': 64.08955009890927}

# data가 많을수록 좋다의 의미:
1. 문제가 있는 데이터는 없애도 된다. (outlier같은거)
2. 적당한 양을 학습시키고, 나머지는 test data로 쓰면 되니까  
다다익선